### ISSUE TITLE
Unexpected keyword argument 'out_types' running test_topk.py

### ISSUE THREAD

**Issue by gwtaylor** [uoguelph-mlrg/PROTAX-GPU#16]
Running this with CUDA and CUDNN installed from pip wheels (`cuda11`) I get an `unexpected keyword argument 'out_types'` when running `test_topk.py`.

Versions are:
```
jaxlib==0.4.20+cuda11.cudnn86
nvidia_cuda_runtime_cu11-11.8.89-py3-none-manylinux1_x86_64.whl
nvidia_cudnn_cu11-8.9.6.50-py3-none-manylinux1_x86_64.whl
```

```bash
(protax-gpu) gwtaylor@gpu068:~/projects/PROTAX-GPU$ python tests/test_topk.py
Traceback (most recent call last):
  File "/fs01/home/gwtaylor/projects/PROTAX-GPU/tests/test_topk.py", line 115, in <module>
    test_topk_small()
  File "/fs01/home/gwtaylor/projects/PROTAX-GPU/tests/test_topk.py", line 80, in test_topk_small
    res = knn(N2S_SMALL.indptr, N2S_SMALL.indices, N2S_SMALL.data, N).block_until_ready()
  File "/scratch/ssd001/home/gwtaylor/mambaforge/envs/protax-gpu/lib/python3.12/site-packages/knn_jax/knn_jax.py", line 40, in knn
    return _knn_prim.bind(indptr, indices, matdat, res)
jax._src.source_info_util.JaxStackTraceBeforeTransformation: TypeError: custom_call() got an unexpected keyword argument 'out_types'

The preceding stack trace is the source of the JAX operation that, once transformed by JAX, triggered the following exception.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/fs01/home/gwtaylor/projects/PROTAX-GPU/tests/test_topk.py", line 115, in <module>
    test_topk_small()
  File "/fs01/home/gwtaylor/projects/PROTAX-GPU/tests/test_topk.py", line 80, in test_topk_small
    res = knn(N2S_SMALL.indptr, N2S_SMALL.indices, N2S_SMALL.data, N).block_until_ready()
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/ssd001/home/gwtaylor/mambaforge/envs/protax-gpu/lib/python3.12/site-packages/knn_jax/knn_jax.py", line 40, in knn
    return _knn_prim.bind(indptr, indices, matdat, res)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/ssd001/home/gwtaylor/mambaforge/envs/protax-gpu/lib/python3.12/site-packages/jax/_src/core.py", line 385, in bind
    return self.bind_with_trace(find_top_trace(args), args, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/ssd001/home/gwtaylor/mambaforge/envs/protax-gpu/lib/python3.12/site-packages/jax/_src/core.py", line 388, in bind_with_trace
    out = trace.process_primitive(self, map(trace.full_raise, args), params)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/ssd001/home/gwtaylor/mambaforge/envs/protax-gpu/lib/python3.12/site-packages/jax/_src/core.py", line 868, in process_primitive
    return primitive.impl(*tracers, **params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/ssd001/home/gwtaylor/mambaforge/envs/protax-gpu/lib/python3.12/site-packages/jax/_src/dispatch.py", line 128, in apply_primitive
    compiled_fun = xla_primitive_callable(
                   ^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/ssd001/home/gwtaylor/mambaforge/envs/protax-gpu/lib/python3.12/site-packages/jax/_src/util.py", line 284, in wrapper
    return cached(config.config._trace_context(), *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/ssd001/home/gwtaylor/mambaforge/envs/protax-gpu/lib/python3.12/site-packages/jax/_src/util.py", line 277, in cached
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/scratch/ssd001/home/gwtaylor/mambaforge/envs/protax-gpu/lib/python3.12/site-packages/jax/_src/dispatch.py", line 157, in xla_primitive_callable
    computation = sharded_lowering(
                  ^^^^^^^^^^^^^^^^^
  File "/scratch/ssd001/home/gwtaylor/mambaforge/envs/protax-gpu/lib/python3.12/site-packages/jax/_src/dispatch.py", line 188, in sharded_lowering
    return pxla.lower_sharding_computation(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/ssd001/home/gwtaylor/mambaforge/envs/protax-gpu/lib/python3.12/site-packages/jax/_src/profiler.py", line 340, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/ssd001/home/gwtaylor/mambaforge/envs/protax-gpu/lib/python3.12/site-packages/jax/_src/interpreters/pxla.py", line 2031, in lower_sharding_computation
    nreps, tuple_args, shape_poly_state) = _cached_lowering_to_hlo(
                                           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/ssd001/home/gwtaylor/mambaforge/envs/protax-gpu/lib/python3.12/site-packages/jax/_src/interpreters/pxla.py", line 1832, in _cached_lowering_to_hlo
    lowering_result = mlir.lower_jaxpr_to_module(
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/ssd001/home/gwtaylor/mambaforge/envs/protax-gpu/lib/python3.12/site-packages/jax/_src/interpreters/mlir.py", line 806, in lower_jaxpr_to_module
    lower_jaxpr_to_fun(
  File "/scratch/ssd001/home/gwtaylor/mambaforge/envs/protax-gpu/lib/python3.12/site-packages/jax/_src/interpreters/mlir.py", line 1213, in lower_jaxpr_to_fun
    out_vals, tokens_out = jaxpr_subcomp(
                           ^^^^^^^^^^^^^^
  File "/scratch/ssd001/home/gwtaylor/mambaforge/envs/protax-gpu/lib/python3.12/site-packages/jax/_src/interpreters/mlir.py", line 1431, in jaxpr_subcomp
    ans = rule(rule_ctx, *rule_inputs, **eqn.params)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/ssd001/home/gwtaylor/mambaforge/envs/protax-gpu/lib/python3.12/site-packages/knn_jax/knn_jax.py", line 93, in _knn_lowering
    out = custom_call(
          ^^^^^^^^^^^^
TypeError: custom_call() got an unexpected keyword argument 'out_types'
```

**Comment by Cleanliness** [comment]
The function signature of `hlo_helpers.custom_call` seems to have changed in jaxlib 0.4.20. [(see here)](https://github.com/google/jax/blob/main/jaxlib/hlo_helpers.py#L134), which is causing the error. For reference, here is what `custom_call` looks like in the jaxlib version (0.4.14) PROTAX-GPU is using:

```
def custom_call(
    call_target_name: Union[str, bytes],
    out_types: Sequence[ir.Type],
    operands: Sequence[ir.Value],
    operand_layouts: Optional[Sequence[Sequence[int]]] = None,
    result_layouts: Optional[Sequence[Sequence[int]]] = None,
    backend_config: str = "",
    has_side_effect: bool = False,
    api_version: int = 2,
    operand_output_aliases: dict[int, int] = {},
    result_shapes: Optional[Sequence[ir.Value]] = None,
) -> Sequence[ir.Value]:
```

TLDR: `out_types` was changed to `result_types` in newer versions of `jaxlib`

**Comment by gwtaylor** [comment]
Ah, ok thanks for that! It looks like JAX 0.4.16 is the first version with `out_types` changed to `result_types`. Falling back to JAX 0.4.14 I no longer get that error. 

I'm able to fully run `test_topk.py`:

```bash
(protax-gpu) gwtaylor@gpu087:~/projects/PROTAX-GPU$ python tests/test_topk.py
test_topk_small: 0.21418046951293945s
failed test_topk_platforms:
test_topk_small: 0.03323078155517578s
passed test_topk_small
test_topk_1K: 0.13089370727539062s
passed test_topk_1k
test_topk_40K: 0.12756586074829102s
passed test_topk_40k
```

Noting that it's a little work for people to install the older JAX, I think we should either update the code to use `result_types` and require JAX>= 0.4.16 or introduce some conditional logic that checks jaxlib version and changes the function call accordingly. What do you think?
